src/components/frontend-engine/yup/helper.ts

```diff
@@ -0,0 +1,177 @@
+	 * @param schemaRules An array of validation rules to be mapped against validation type (i.e. a string schema might contain { maxLength: 255 })
+	 * @returns yupSchema with added constraints and validations
+	 */
+	const mapRules = (yupSchema: Yup.AnySchema, schemaRules: IYupCombinedRule[]): Yup.AnySchema => {
+		schemaRules.forEach((rule) => {
+			const ruleKey = Object.keys(rule).filter((k) =>
+				YUP_CONDITIONS.includes(k as TYupCondition)
+			)?.[0] as TYupCondition;
+
+			switch (true) {
+				case rule.required:
+					yupSchema = yupSchema.required(rule.errorMessage || "This field is required");
+					break;
+				case !!rule.email:
+				case !!rule.url:
+				case !!rule.uuid:
+				case !!rule.positive:
+				case !!rule.negative:
+				case !!rule.integer:
+					yupSchema = (yupSchema as unknown)[ruleKey](rule.errorMessage);
+					break;
+				case rule.length > 0:
+				case rule.min > 0:
+				case rule.max > 0:
+				case !!rule.filled:
+				case !!rule.empty:
+				case !!rule.equals:
+				case !!rule.notEquals:
+				case !!rule.includes:
+				case !!rule.excludes:
+				case !!rule.uinfin:
+					yupSchema = (yupSchema as unknown)[ruleKey](rule[ruleKey], rule.errorMessage);
+					break;
+				case !!rule.matches:
+					{
+						const matches = rule.matches.match(/\/(.*)\/([a-z]+)?/);
+						yupSchema = (yupSchema as Yup.StringSchema).matches(
+							new RegExp(matches[1], matches[2]),
+							rule.errorMessage
+						);
+					}
+					break;
+				case !!rule.when:
+					{
+						Object.keys(rule.when).forEach((fieldId) => {
+							const isRule = rule.when[fieldId].is;
+							const thenRule = mapRules(
+								mapSchemaType(yupSchema.type as TYupSchemaType),
+								rule.when[fieldId].then
+							);
+							const otherwiseRule =
+								rule.when[fieldId].otherwise &&
+								mapRules(mapSchemaType(yupSchema.type as TYupSchemaType), rule.when[fieldId].otherwise);
+
+							if (Array.isArray(isRule) && (isRule as unknown[]).every((r) => typeof r === "object")) {
+								yupSchema = yupSchema.when(fieldId, (value: unknown, fieldYupSchema: Yup.AnySchema) => {
+									const localYupSchema = mapRules(
+										fieldYupSchema.clone(),
+										isRule as IYupConditionalValidationRule[]
+									);
+									let fulfilled = false;
+									try {
+										localYupSchema.validateSync(value);
+										fulfilled = true;
+									} catch (error) {}
+
+									return fulfilled ? thenRule : otherwiseRule;
+								});
+							} else {
+								yupSchema = yupSchema.when(fieldId, {
+									is: isRule,
+									then: thenRule,
+									otherwise: otherwiseRule,
+								});
+							}
+						});
+					}
+					break;
+			}
+		});
+
+		return yupSchema;
+	};
+
+	/**
+	 * Declare custom Yup schema condition
+	 * @param type The schema type
+	 * @param name Name of the condition
+	 * @param fn Validation function, it must return a boolean
+	 */
+	export const addCondition = (
```

Comment:
note: this is a new function to add custom validation internally (i.e. by us)
more works need to be done for devs to add custom validation outside `FrontendEngine`